tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,339 @@
+                  [parent_graph_name][child_remote_op_name])
+
+  mapping = name_mapping.items()
+  # Calling _copy_tensor multiple times may introduce a burden, since
+  # _copy_tensor invokes a deepcopy on element.
+  for child_graph_placeholder_name, parent_graph_input_name in mapping:
+
+    step_name = ("PrepareInput[Graph_%s][Input_%s]" %
+                 (child_remote_op_name, child_graph_placeholder_name))
+    pcoll = pcoll | step_name >> beam.Map(
+        _copy_tensor,
+        parent_remote_op_name,
+        _import_tensor_name(parent_graph_input_name),
+        child_remote_op_name,
+        _import_tensor_name(child_graph_placeholder_name))
+
+  return pcoll
+
+
+def _copy_tensor(  # pylint: disable=invalid-name
+    element: Dict[Text, Dict[Text, tf.Tensor]],
+    old_graph: Text, old_tensor_name: Text,
+    new_graph: Text, new_tensor_name: Text
+) -> Dict[Text, Dict[Text, tf.Tensor]]:
+  element = copy.deepcopy(element)
+  if new_graph not in element:
+    element[new_graph] = {}
+  element[new_graph][new_tensor_name] = element[old_graph][old_tensor_name]
+  return element
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(Dict[Text, Dict[Text, tf.Tensor]])
+@beam.typehints.with_output_types(Dict[Text, Dict[Text, tf.Tensor]])
+def _ExtractRemoteGraphOutput(  # pylint: disable=invalid-name
+    pcoll: beam.pvalue.PCollection,
+    parent_remote_op_name: Text,
+    child_remote_op_name: Text,
+    remote_op_name_to_graph_name: Mapping[Text, Text],
+    graph_name_to_specs: Mapping[Text, List[execution_spec.ExecutionSpec]],
+) -> beam.pvalue.PCollection:
+  """Extracts remote graph output.
+
+  After finish executing a remote graph, we need to collect its output.
+  We first find the output name of the remote graph, then we copy the
+  output of the remote graph to its parent graph. Finally, we clear the
+  intermediate results of the remote graph.
+
+  Note we assumed that each node has only one output, which also applies
+  to remote op. This means that a remote graph can only have one output.
+
+  Args:
+    pcoll: A PCollection of child graph results. Each element is a dictionary
+           from remote op names to a dictionary from tensor names to tensors.
+           Here, element[child_remote_op_name] stores graph inputs,
+           intermediate results, and graph output.
+    parent_remote_op_name: The remote op name of the parent graph.
+    child_remote_op_name: The remote op name of the child graph.
+    remote_op_name_to_graph_name:
+      A mapping from remote op names to graph names.
+    graph_name_to_specs: A mapping from graph names to a list of
+                         ExecutionSpecs.
+
+  Returns:
+    A PCollection of child graph output in parent graph. Each element is a
+    dictionary from remote op names to a dictionary from tensor names to
+    tensors. Here, element[parent_remote_op_name] contains the output from
+    the child graph, and element[child_remote_op_name] becomes empty.
```

Comment:
"is deleted" ("becomes empty" is ambiguous because it might mean merely deleting all items in the dictionary, rather than deleting the whole dictionary)